### PR TITLE
[trace-sdk-sampler] Differentiate between root and child span creation in trace sampling spec

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Allow SDKs to bypass span creation for children.

## Reasoning

The .NET SDK works like this today: If the sampler returns `DROP` we always create a span for the root. For children, we don't create a span. This takes .NET out of compliance with the spec (as-is) but we have been resistant to change this behavior because it is expensive to create these spans (memory, cpu, gc pressure, etc.) for benefits which seem unclear. If the purpose is just to give a valid non-empty SpanId to logs (and friends) then the root creation behavior seems to accomplish the goal more cheaply.

Also, .NET also already has a mechanism which is more or less #3867 where users can drop a source/instrumentation scope. Really everything is no-op by default users have to explicitly enable the sources/instrumentation scopes they care about. Even if .NET was spec-compliant w.r.t. the span creation traces could still have gaps for things users haven't switched on. This isn't something .NET SDK can change either, as it comes from the underlying .NET platform.

What this PR is seeking to do is officially bless what .NET has done in order to bring it into spec-compliance so we may resolve things like https://github.com/open-telemetry/opentelemetry-dotnet/issues/3290.

## TODOs

* [ ] [`CHANGELOG.md`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/CHANGELOG.md) file updated for non-trivial changes

### CODE REVIEW

**Comment by pellared** [comment]
Have you checked what other SDKs are doing in such scenarios? It is a user-facing behavior so it would be good if all SDK would do the same.

File: specification/trace/sdk.md [hunk 1]
```diff
@@ -232,9 +232,12 @@ When asked to create a Span, the SDK MUST act as if doing the following in order
    (Note that the [built-in `ParentBasedSampler`](#parentbased) can be used to
    use the sampling decision of the parent,
    translating a set SampledFlag to RECORD and an unset one to DROP).
-3. Generate a new span ID for the `Span`, independently of the sampling decision.
-   This is done so other components (such as logs or exception handling) can rely on
-   a unique span ID, even if the `Span` is a non-recording instance.
+3. If there isn't currently a span (the span being created will become the root
```
Comment (cijothomas) [inline specification/trace/sdk.md:235]: @Oberon00 to share early comments. (the PR is still draft)

Comment (cijothomas) [inline specification/trace/sdk.md:235]:
```suggestion
3. If there isn't currently a span (i.e the span being created will become the root
```

File: specification/trace/sdk.md [hunk 2]
```diff
@@ -232,9 +232,12 @@ When asked to create a Span, the SDK MUST act as if doing the following in order
...
+   span), SDKs MUST generate a new span ID for the `Span`, independently of the
+   sampling decision. This is done so other components (such as logs or
+   exception handling) can rely on a unique span ID, even if the `Span` is a
+   non-recording instance. For children, SDKs MAY skip creation of spans if
```
Comment (cijothomas) [inline specification/trace/sdk.md:239]:
```suggestion
   non-recording instance. For non-root spans, SDKs MAY skip creation of spans if
```

File: specification/trace/sdk.md [hunk 3]
```diff
@@ -232,9 +232,12 @@ When asked to create a Span, the SDK MUST act as if doing the following in order
...
+   `ShouldSample` results in a non-recording instance (`DROP`).
```
Comment (cijothomas) [inline specification/trace/sdk.md:240]:
```suggestion
   `ShouldSample` results in a non-recording instance (`DROP`). This is done to optimize the performance by avoiding creating a span when unnecessary.
```

**Comment by CodeBlanch** [comment]
@cijothomas Are you able to reopen this? Doesn't seem like I can do it.

**Comment by Oberon00** [comment]
While the change seems sensible (without thinking very deep about it), it is also a breaking change to a stable specification part.

**Comment by cijothomas** [comment]
> While the change seems sensible (without thinking very deep about it), it is also a breaking change to a stable specification part.

I think the proposed wording in this PR is written is such a way so as to be back-compatible, very similar to the OTLP Exporter defaulting to http wording from diff. section of spec.

> If no configuration is provided the default transport SHOULD be http/protobuf unless SDKs have good reasons to choose grpc as the default (e.g. for backward compatibility reasons when grpc was already the default in a stable SDK release).

**Comment by CodeBlanch** [comment]
@Oberon00 

> it is also a breaking change to a stable specification part

@jack-berg Had the same feedback for me when I first shared this idea with him. We are in no way recommending SDKs change their behavior. What we want to do is retroactively bless the direction .NET has already taken.

I just pushed an update which makes this explicit. Feels very unnecessary to me, but if it helps makes this clear doesn't seem like a problem to have it.

**Pull request closed without merging**